projects/packages/partner/src/class-partner-coupon.php

```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
+	 * errors.
 	 */
-	public function purge_coupon() {
+	public function maybe_purge_coupon() {
+		// Only run coupon checks on Jetpack admin pages.
+		// The "admin-ui" package is responsible for registering the Jetpack admin
+		// page for all Jetpack plugins and has hardcoded the settings page to be
+		// "jetpack", so we shouldn't need to allow for dynamic/custom values.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! isset( $_GET['page'] ) || ! 'jetpack' === $_GET['page'] ) {
+			return;
+		}
+
+		if ( $this->maybe_purge_coupon_by_added_date() ) {
+			return;
+		}
+
+		$last_check = Jetpack_Options::get_raw_option( self::$last_check_option, 0 );
+
+		// Limit checks to happen every minute instead of on all page requests.
+		if ( $last_check && time() > $last_check + MINUTE_IN_SECONDS ) {
+			return;
+		}
+
+		Jetpack_Options::update_raw_option( self::$last_check_option, time() );
+
+		$this->maybe_purge_coupon_by_availability_check();
+	}
+
+	/**
+	 * Purge coupon based on local added date.
+	 *
+	 * We automatically remove the coupon after a month to "self-heal" if
+	 * something in the claim process has broken with the site.
+	 *
+	 * @return bool Return whether we should skip further purge checks.
+	 */
+	protected function maybe_purge_coupon_by_added_date() {
 		$date = Jetpack_Options::get_option( self::$added_option, '' );
 
 		if ( empty( $date ) ) {
-			return;
+			return true;
 		}
 
 		$expire_date = strtotime( '+30 days', $date );
 		$today       = time();
 
 		if ( $today >= $expire_date ) {
-			Jetpack_Options::delete_option(
-				array(
-					self::$coupon_option,
-					self::$added_option,
-				)
-			);
+			$this->delete_coupon_data();
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Purge coupon based on availability check.
+	 *
+	 * @param string|null $connection_client The connection client to remotely look up availability.
+	 * @return bool Return whether we deleted coupon data.
+	 */
+	protected function maybe_purge_coupon_by_availability_check( $connection_client = null ) {
+		// Allow for client dependency injection to allow for mock client during unit testing.
+		if (
+			empty( $connection_client ) ||
+			! class_exists( $connection_client ) ||
+			! method_exists( $connection_client, 'wpcom_json_api_request_as_blog' )
+		) {
+			$connection_client = Connection_Client::class;
+		}
+
+		$jetpack_id = Jetpack_Options::get_option( 'id', false );
+		if ( ! $jetpack_id ) {
+			return false;
+		}
+
+		$coupon   = self::get_coupon();
+		$response = $connection_client::wpcom_json_api_request_as_blog(
+			sprintf(
+				'/sites/%d/jetpack-partner/coupon/v1/site/coupon?coupon_code=%s',
+				$jetpack_id,
+				$coupon['coupon_code']
+			),
+			2,
+			array( 'method' => 'GET' ),
+			null,
+			'wpcom'
+		);
+
+		$body = json_decode( wp_remote_retrieve_body( $response ), true );
+
+		if (
+			200 === wp_remote_retrieve_response_code( $response ) &&
+			isset( $body['available'] ) &&
```

Comment:
This seems like it can raise an error if the decoded value is not an array - isset might avoid it, but it won't hurt to add an is_array() check as well. Leaving this entirely up to you.